### PR TITLE
Patch for debugger notebooks with profiling

### PR DESCRIPTION
**Pull request by ndodda-amazon**
*Issue #, if available:*

*Description of changes:*
Various necessary bug fixes in the notebooks.
- Upgrade to the latest sagemaker SDK in all of the notebooks
- Install smdebug in all notebooks where it used
- Update the profiler config for TF 2.3.1 jobs to specify the local path for framework profiling
- Use the image URI for the DLC image directly instead of relying on framework and python version in the estimator. Dynamically determine the region to complete the image URI.
- Swap the values for `instance_type` and `instance_count` in the bottleneck notebooks.
- Remove the array wrapped around the rules specified in the estimator.
- Initialize the system metrics reader wherever it is used.
- Lower the batch size in the training job for the callback notebook.

After applying the above fixes, I tested in CMH (us-east-2). All of the notebooks execute correctly. All of the training jobs from the notebooks pass except for the dataset bottleneck notebook, which fails due to an obscure error. For now, we will exclude this notebook and add it back again once we have fixed the notebook.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb [hunk 1]
```diff
@@ -9,14 +9,37 @@
     "This notebook will walk you through creating a PyTorch training job with the SageMaker Debugger profiling feature enabled. It will create a multi GPU multi node training.\n"
```
Comment (mchoi8739) [inline sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: <p>Change to restart kernel automatically, if you want to put pip install.</p><pre class="ql-syntax" spellcheck="false">import sys
import IPython
install_needed = True  # should only be True once
if install_needed:
    print("installing deps and restarting kernel")
    !{sys.executable} -m pip install -U sagemaker
    !{sys.executable} -m pip pip install -U smdebug
    IPython.Application.instance().kernel.do_shutdown(True)
</pre><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1844/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (mchoi8739) [inline sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: <p>To enable profiling, you need to pass the Debugger profiling configuration (<code>profiler_config</code>), a list of Debugger rules (<code>rules</code>), and the image URI (<code>image_uri</code>) to the estimator. Debugger enables monitoring and profiling while the SageMaker estimator requests a training job.</p><p> </p><p> </p><p>^ please apply this throughout the other notebooks too.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1844/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-single-node.ipynb [hunk 2]
```diff
@@ -9,14 +9,37 @@
     "This notebook will walk you through creating a PyTorch training job with the SageMaker Debugger profiling feature enabled. It will create a multi GPU single node training using Horovod.\n"
```
Comment (mchoi8739) [inline sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-single-node.ipynb:9]: <pre class="ql-syntax" spellcheck="false">import sys
import IPython
install_needed = True  # should only be True once
if install_needed:
    print("installing deps and restarting kernel")
    !{sys.executable} -m pip install -U sagemaker
    !{sys.executable} -m pip pip install -U smdebug
    IPython.Application.instance().kernel.do_shutdown(True)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1844/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-single-gpu-single-node.ipynb [hunk 3]
```diff
@@ -9,14 +9,37 @@
     "This notebook will walk you through creating a PyTorch training job with the SageMaker Debugger profiling feature enabled. It will create a single GPU single node training.\n"
```
Comment (mchoi8739) [inline sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-single-gpu-single-node.ipynb:9]: <pre class="ql-syntax" spellcheck="false">import sys
import IPython
install_needed = True  # should only be True once
if install_needed:
    print("installing deps and restarting kernel")
    !{sys.executable} -m pip install -U sagemaker
    !{sys.executable} -m pip pip install -U smdebug
    IPython.Application.instance().kernel.do_shutdown(True)
</pre><p> </p><p>would you update this throughout the all other notebooks?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1844/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb (hunk 1 above)
Comment (mchoi8739) [inline sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: <p>To use the new Debugger profiling features, ensure that you have the latest versions of SageMaker and SMDebug SDKs installed. The following cell updates the libraries and restart the kernel to apply the updates to your working environment.</p><p> </p><p> </p><p>^ please change this throughout the notebooks</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1844/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ndodda-amazon) [reply sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: Done.

Comment (ndodda-amazon) [reply sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: Done.

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-single-node.ipynb (hunk 2 above)
Comment (ndodda-amazon) [reply sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-single-node.ipynb:9]: Done.

File: sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb (hunk 1 above)
Comment (ndodda-amazon) [reply sagemaker-debugger/pytorch_profiling/pt-resnet-profiling-multi-gpu-multi-node.ipynb:9]: Done.

**Review by mchoi8739 — APPROVED** [review]

**Review by ajaykarpur — APPROVED** [review]

**Pull request merged**